File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RankingItemView.java
package com.garpr.android.views;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.util.AttributeSet;
import android.view.View;
import android.widget.TextView;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.activities.PlayerActivity;
import com.garpr.android.adapters.BaseAdapterView;
import com.garpr.android.misc.FavoritePlayersManager;
import com.garpr.android.misc.RegionManager;
import com.garpr.android.models.Ranking;

import java.text.NumberFormat;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class RankingItemView extends IdentityFrameLayout implements BaseAdapterView<Ranking>,
        View.OnClickListener, View.OnLongClickListener {

    private NumberFormat mNumberFormat;
    private Ranking mContent;

    @Inject
    FavoritePlayersManager mFavoritePlayersManager;

    @Inject
    RegionManager mRegionManager;

    @Nullable
    @BindView(R.id.previousRankView)
    PreviousRankView mPreviousRankView;

    @BindView(R.id.tvName)
    TextView mName;

    @BindView(R.id.tvRank)
    TextView mRank;

    @BindView(R.id.tvRating)
    TextView mRating;


    public RankingItemView(@NonNull final Context context, @Nullable final AttributeSet attrs) {
        super(context, attrs);
    }

    public RankingItemView(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public RankingItemView(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr, @StyleRes final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected String getIdentityId() {
        return mContent.getId();
    }

    @Override
    protected void identityIsSomeoneElse() {
        super.identityIsSomeoneElse();
        styleTextViewForSomeoneElse(mName);
    }

    @Override
    protected void identityIsUser() {
        super.identityIsUser();
        styleTextViewForUser(mName);
    }

    @Override
    public void onClick(final View v) {
        final Context context = getContext();
        context.startActivity(PlayerActivity.Companion.getLaunchIntent(context, mContent,
                mRegionManager.getRegion(context)));
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        ButterKnife.bind(this);
        setOnClickListener(this);
        setOnLongClickListener(this);
        mNumberFormat = NumberFormat.getNumberInstance();

        if (isInEditMode()) {
            return;
        }

        App.get().getAppComponent().inject(this);
    }

    @Override
    public boolean onLongClick(final View v) {
        return mFavoritePlayersManager.showAddOrRemovePlayerDialog(getContext(),
                mContent.getPlayer(), mRegionManager.getRegion(getContext()));
    }

    @Override
    public void setContent(final Ranking content) {
        mContent = content;

        if (mPreviousRankView != null) {
            mPreviousRankView.setContent(mContent);
        }

        mRank.setText(mNumberFormat.format(mContent.getRank()));
        mName.setText(mContent.getName());
        mRating.setText(mContent.getRatingTruncated());

        refreshIdentity();
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RankingItemView.kt
package com.garpr.android.views

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.core.content.ContextCompat
import com.garpr.android.R
import com.garpr.android.activities.PlayerActivity
import com.garpr.android.adapters.BaseAdapterView
import com.garpr.android.extensions.appComponent
import com.garpr.android.extensions.clear
import com.garpr.android.extensions.setTintedImageResource
import com.garpr.android.managers.FavoritePlayersManager
import com.garpr.android.managers.RegionManager
import com.garpr.android.misc.MiscUtils
import com.garpr.android.misc.PreviousRankUtils
import com.garpr.android.misc.Timber
import com.garpr.android.models.RankedPlayer
import kotlinx.android.synthetic.main.item_ranking.view.*
import java.text.NumberFormat
import javax.inject.Inject

class RankingItemView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : IdentityConstraintLayout(context, attrs), BaseAdapterView<RankedPlayer>,
        View.OnClickListener, View.OnLongClickListener {

    private val numberFormat = NumberFormat.getIntegerInstance()

    @Inject
    protected lateinit var favoritePlayersManager: FavoritePlayersManager

    @Inject
    protected lateinit var previousRankUtils: PreviousRankUtils

    @Inject
    protected lateinit var regionManager: RegionManager

    @Inject
    protected lateinit var timber: Timber


    override fun clear() {
        super.clear()
        previousRankView.clear()
        rank.clear()
        name.clear()
        rating.clear()
    }

    override fun identityIsSomeoneElse() {
        super.identityIsSomeoneElse()
        styleTextViewForSomeoneElse(name)
    }

    override fun identityIsUser() {
        super.identityIsUser()
        styleTextViewForUser(name)
    }

    override fun onClick(v: View) {
        val identity = this.identity ?: return
        context.startActivity(PlayerActivity.getLaunchIntent(context, identity,
                regionManager.getRegion(context)))
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        if (!isInEditMode) {
            appComponent.inject(this)
        }

        setOnClickListener(this)
        setOnLongClickListener(this)
    }

    override fun onLongClick(v: View): Boolean {
        return favoritePlayersManager.showAddOrRemovePlayerDialog(context, identity,
                regionManager.getRegion(context))
    }

    override fun refresh() {
        super.refresh()

        val player = identity as? RankedPlayer

        if (player == null) {
            clear()
            return
        }

        val rankInfo = previousRankUtils.getRankInfo(player)

        if (rankInfo == null) {
            previousRankView.clear()
            previousRankView.visibility = View.GONE
        } else {
            @DrawableRes val drawableResId: Int
            @ColorRes val tintResId: Int

            when (rankInfo) {
                PreviousRankUtils.Info.DECREASE -> {
                    drawableResId = R.drawable.ic_arrow_downward_white_18dp
                    tintResId = R.color.lose
                }

                PreviousRankUtils.Info.INCREASE -> {
                    drawableResId = R.drawable.ic_arrow_upward_white_18dp
                    tintResId = R.color.win
                }
            }

            previousRankView.setTintedImageResource(drawableResId,
                    ContextCompat.getColor(context, tintResId))
            previousRankView.visibility = View.VISIBLE
        }

        rank.text = numberFormat.format(player.rank)
        name.text = player.name
        rating.text = MiscUtils.truncateFloat(player.rating)
    }

    override fun setContent(content: RankedPlayer) {
        identity = content
        refresh()
    }

}
